File: trunk/dotnet/picocontainer/src/Extras/ImplementationHidingComponentAdapterFactory.cs
/*****************************************************************************
 * Copyright (C) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the license.txt file.                                                     *
 *                                                                           *
 * Idea by Rachel Davies, Original code by Aslak Hellesoy and Paul Hammant   *
 * C# port by Maarten Grootendorst                                           *
 *****************************************************************************/

using System;
using System.Collections;
using PicoContainer;
using PicoContainer.Defaults;
using PicoContainer.Lifecycle;


namespace PicoContainer.Extras {
  public class ImplementationHidingComponentAdapterFactory : DecoratingComponentAdapterFactory {
    private static InterfaceFinder interfaceFinder = new InterfaceFinder();

    public ImplementationHidingComponentAdapterFactory() : base (new DefaultComponentAdapterFactory()) {
    }

    public ImplementationHidingComponentAdapterFactory(ComponentAdapterFactory theDelegate) : base (theDelegate) {
    }

    public override ComponentAdapter CreateComponentAdapter(object componentKey,
      Type componentImplementation,
      Parameter[] parameters) {
      return new Adapter(base.CreateComponentAdapter(componentKey, componentImplementation, parameters));
    }

    public void HotSwap(Type type) {
    }

    public class Adapter : DecoratingComponentAdapter {
      public Adapter(ComponentAdapter theDelegate) : base (theDelegate) {
      }

      public override Object GetComponentInstance(MutablePicoContainer picoContainer) {
        Object component = base.GetComponentInstance(picoContainer);
        Type[] interfaces = interfaceFinder.GetInterfaces(component);
        if(interfaces.Length == 0) {
          throw new Exception("Can't hide implementation");
          //          throw new PicoIntrospectionException("Can't hide implementation for " + component.getClass().getName() + ". It doesn't implement any interfaces.");
        }
        Type t = DefaultComponentMulticasterFactory.GetAggregatingInterface(interfaces);
        ArrayList list = new ArrayList();
        list.Add(component);
        return new AggregatingProxy(t ,list).GetTransparentProxy();
      }
    }
  }
}

File: java/picocontainer/src/java/org/picocontainer/defaults/ImplementationHidingComponentAdapterFactory.java
/*****************************************************************************
 * Copyright (c) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the LICENSE.txt file.                                                     *
 *                                                                           *
 * Idea by Rachel Davies, Original code by Aslak Hellesoy and Paul Hammant   *
 *****************************************************************************/

package org.picocontainer.defaults;

import org.picocontainer.ComponentAdapter;
import org.picocontainer.Parameter;
import org.picocontainer.PicoIntrospectionException;

/**
 * @author Paul Hammant
 * @author Aslak Helles&oslash;y
 * @version $Revision$
 */
public class ImplementationHidingComponentAdapterFactory extends DecoratingComponentAdapterFactory {
    private final boolean strict;

    public ImplementationHidingComponentAdapterFactory() {
        this(new DefaultComponentAdapterFactory(), true);
    }

    public ImplementationHidingComponentAdapterFactory(ComponentAdapterFactory delegate) {
        this(delegate, true);
    }

    public ImplementationHidingComponentAdapterFactory(ComponentAdapterFactory delegate, boolean strict) {
        super(delegate);
        this.strict = strict;
    }

    public ComponentAdapter createComponentAdapter(Object componentKey,
                                                   Class componentImplementation,
                                                   Parameter[] parameters)
            throws PicoIntrospectionException, AssignabilityRegistrationException, NotConcreteRegistrationException {
        return new ImplementationHidingComponentAdapter(super.createComponentAdapter(componentKey, componentImplementation, parameters), strict);
    }
}
